### PR TITLE
Pending update directly requests for stable if it hits stable karma

### LINKED ISSUE #632: should push to stable immediately if will be automatically set to request stable at time of push to testing
Example:
https://bodhi.fedoraproject.org/updates/FEDORA-2015-16569

This update was pushed to testing at 2015-09-27 00:42:54
It was then automatically requested for stable based upon karma 42 seconds later.
Then the package had to wait 15 hours to be pushed to stable, even though it already met the karma requirements for automatic request for stable before the push to testing.

Instead of (or in addition to?) pushing it to testing, it should have been automatically pushed to stable at the time it pushed to testing, rather than having to wait until the next time updates get pushed, which in this case was 15 hours later (in some cases the wait seems to be longer still).

Probably the best way to approach this is for Bodhi to automatically change the request from testing to stable if the Karma requirements for automatic request for stable have been reached before the package makes it to testing.

Timeliness of updates is important for security, for example. In this case, the sqlite package was holding up for several days a security update for firefox which depended upon it (https://bodhi.fedoraproject.org/updates/FEDORA-2015-16455). (Actually, it looks like this firefox package is another example where the push to stable was delayed for the same reason.)

### PR DESCRIPTION
**Pull request by trishnaguha**
Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>
Fixes #632 
- Requests for stable if pending update hits stable karma on Autopush enabled/disabled.

### CODE REVIEW

File: bodhi/server/models.py [hunk 1]
```diff
@@ -2092,10 +2092,10 @@ def check_karma_thresholds(self, db, agent):
             log.debug('%s locked. Ignoring karma thresholds.' % self.title)
             raise LockedUpdateException
         # Return if the status of the update is not in testing
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: This comment should be updated.

File: bodhi/server/models.py [hunk 2]
```diff
@@ -2092,10 +2092,10 @@ def check_karma_thresholds(self, db, agent):
...
-        if self.status != UpdateStatus.testing:
+        if self.status not in (UpdateStatus.testing, UpdateStatus.pending):
             return
         # If an update receives negative karma disable autopush
-        if self.autokarma and self._composite_karma[1] != 0:
+        if self.autokarma and self._composite_karma[1] != 0 and self.status is UpdateStatus.testing:
```
Comment (bowlofeggs) [inline bodhi/server/models.py:2306]: Why would we only want to disable autokarma for testing updates? It seems to me that we should treat the pending updates the same way.

File: bodhi/server/models.py [hunk 3]
```diff
@@ -2115,7 +2115,8 @@ def check_karma_thresholds(self, db, agent):
                 log.info((
                     "%s update has reached the stable karma threshold and can be pushed to "
                     "stable now if the maintainer wishes") % self.title)
-        elif self.unstable_karma and self.karma <= self.unstable_karma:
+        elif self.unstable_karma and self.karma <= self.unstable_karma \
+                and self.status is UpdateStatus.testing:
```
Comment (bowlofeggs) [inline bodhi/server/models.py]: Similarly here - why treat the testing update differently than the pending one?

File: bodhi/tests/server/services/test_updates.py [hunk 4]
```diff
@@ -1866,6 +1866,84 @@ def test_edit_locked_update(self, publish, *args):
...
+
+        self.assertEquals(up.karma, 2)
+        self.assertEquals(up.status, UpdateStatus.pending)
+
+        id = 'bodhi-2.0.0-2.fc17'
+        resp = self.app.get('/updates/%s' % id,
+                            headers={'Accept': 'text/html'})
+        self.assertIn('text/html', resp.headers['Content-Type'])
+        self.assertIn(id, resp)
+        self.assertNotIn('Push to Stable', resp)
```
Comment (bowlofeggs) [inline bodhi/tests/server/services/test_updates.py]: Wouldn't we want ```self.assertIn('Push to Stable', resp)``` here instead?

**Review by bowlofeggs — CHANGES_REQUESTED** [review]
There seem to be 5 failing tests. Here's one of them:

```
======================================================================
FAIL: test_commenting_twice_with_positive_then_negative_karma (bodhi.tests.server.functional.test_comments.TestCommentsService)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/vagrant/bodhi/bodhi/tests/server/functional/test_comments.py", line 193, in test_commenting_twice_with_positive_then_negative_karma
    self.assertEquals(res.json_body['comment']['text'], 'Test')
AssertionError: u'Bodhi is disabling automatic push to stable due to negative karma. The maintainer may push manually if they determine that the issue is not severe.' != 'Test'
```

**Comment by bowlofeggs** [comment]
ok test

File: bodhi/server/models.py (hunk 2 above)
Comment (trishnaguha) [reply bodhi/server/models.py:2306]: @bowlofeggs We would want to disable autokarma **only** for testing updates as this is conflicting with https://github.com/fedora-infra/bodhi/blob/develop/bodhi/server/models.py#L2380 for pending updates.
As we would not like to disable autopush for pending updates on unstable karma, rather we want to send it to `obsolete` state https://github.com/fedora-infra/bodhi/blob/develop/bodhi/tests/server/services/test_updates.py#L2397

File: bodhi/server/models.py [hunk 5]
```diff
@@ -2323,11 +2323,14 @@ def check_karma_thresholds(self, db, agent):
                     "%s update has reached the stable karma threshold and can be pushed to "
                     "stable now if the maintainer wishes") % self.title)
         elif self.unstable_karma and self.karma <= self.unstable_karma:
-            log.info("Automatically unpushing %s" % self.title)
-            self.obsolete(db)
-            notifications.publish(
-                topic='update.karma.threshold.reach',
-                msg=dict(update=self, status='unstable'))
+            if self.status is UpdateStatus.pending and not self.autokarma:
+                pass
```
Comment (trishnaguha) [inline bodhi/server/models.py:2327]: For pending updates with unstable karma (but with disabled autopush), we would not like to send it to `obsolete` state automatically https://github.com/fedora-infra/bodhi/blob/develop/bodhi/tests/server/services/test_updates.py#L2423

**Comment by bowlofeggs** [comment]
test

**Review by bowlofeggs — APPROVED** [review]

**Comment by bowlofeggs** [comment]
I opened https://github.com/fedora-infra/bodhi/pull/1651 to add one more assertion to one of these tests.

**Pull request merged**